libnetwork_test.go

```diff
@@ -0,0 +1,349 @@
+			DockerString(fmt.Sprintf("docker run --link-local-ip 169.254.0.50 %s --net %s -tid --name %s busybox", name, name))
+
+			// Delete container
+			DockerString(fmt.Sprintf("docker rm -f %s", name))
+		})
+
+		// TODO Ensure that  a specific IP isn't possible without a user specified subnet
+		It("creates a container with specific IP", func() {
+			// Create a network with a chosen subnet as this is required to choose an IP
+			name_subnet := fmt.Sprintf("run%d", rand.Uint32())
+			DockerString(fmt.Sprintf("docker network create %s --subnet 192.169.0.0/16 -d calico --ipam-driver calico-ipam", name_subnet))
+			// Create a container that will just sit in the background
+			chosen_ip := "192.169.50.51"
+			DockerString(fmt.Sprintf("docker run --ip %s --net %s -tid --name %s busybox", chosen_ip, name_subnet, name_subnet))
+
+			// Gather information for assertions
+			docker_endpoint := GetDockerEndpoint(name_subnet, name_subnet)
+			ip := docker_endpoint.IPAddress
+			mac := docker_endpoint.MacAddress
+			endpoint_id := docker_endpoint.EndpointID
+			interface_name_subnet := "cali" + endpoint_id[:mathutils.MinInt(11, len(endpoint_id))]
+
+			Expect(ip).Should(Equal(chosen_ip))
+
+			// Check that the endpoint is created in etcd
+			etcd_endpoint := GetEtcdString(fmt.Sprintf("/calico/v1/host/test/workload/libnetwork/libnetwork/endpoint/%s", endpoint_id))
+			Expect(etcd_endpoint).Should(MatchJSON(fmt.Sprintf(
+				`{"state":"active","name":"%s","mac":"%s","profile_ids":["%s"],"ipv4_nets":["%s/32"],"ipv6_nets":[]}`,
+				interface_name_subnet, mac, name_subnet, ip)))
+
+			// Check the interface exists on the Host - it has an autoassigned
+			// mac and ip, so don't check anything!
+			DockerString(fmt.Sprintf("ip addr show %s", interface_name_subnet))
+
+			// Make sure the interface in the container exists and has the  assigned ip and mac
+			container_interface_string := DockerString(fmt.Sprintf("docker exec -i %s ip addr", name_subnet))
+			Expect(container_interface_string).Should(ContainSubstring(ip))
+			Expect(container_interface_string).Should(ContainSubstring(mac))
+
+			// Make sure the container has the routes we expect
+			routes := DockerString(fmt.Sprintf("docker exec -i %s ip route", name_subnet))
+			Expect(routes).Should(Equal("default via 169.254.1.1 dev cali0 \n169.254.1.1 dev cali0"))
+
+			// Delete container and network
+			DockerString(fmt.Sprintf("docker rm -f %s", name_subnet))
+			DockerString(fmt.Sprintf("docker network rm %s", name_subnet))
+		})
+
+	})
+	//docker stop/rm - stop and rm are the same as far as the plugin is concerned
+	// TODO - check that the endpoint is removed from etcd and that the  veth is removed
+})
+
+func GetDockerEndpoint(container, network string) *network.EndpointSettings {
+	os.Setenv("DOCKER_API_VERSION", "1.24")
+	os.Setenv("DOCKER_HOST", "http://localhost:5375")
+	cli, err := dockerclient.NewEnvClient()
+	if err != nil {
+		panic(err)
+	}
+	os.Setenv("DOCKER_HOST", "")
```

Comment:
Should this be done using defer ... or does it need to be done before the inspect?   just wondering if a failed test could end up causing follow on failures.